File: trees/binary_tree_right_side_dfs.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def right_side_view(root)
    right_view = []

    return right_view if root.nil?

    dfs_helper(root, 0, right_view)
    right_view
end

def dfs_helper(node, depth, right_view)
    return if node.nil?

    # If this is the first time we're visiting this depth level,
    # add the current node's value (rightmost at this level)
    if depth == right_view.size
        right_view << node.val
    end

    # Visit right child first, then left child
    # This ensures we see the rightmost nodes first at each level
    dfs_helper(node.right, depth + 1, right_view)
    dfs_helper(node.left, depth + 1, right_view)
end

File: trees/binary_tree_right_side_dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        right_view = []

        if root is None:
            return right_view

        self.dfs(root, 0, right_view)
        return right_view

    def dfs(self, node, depth, right_view):

        if node is None:
            return
        
        if depth == len(right_view):
            right_view.append(node.val)

        self.dfs(node.right, depth + 1, right_view)
        self.dfs(node.left, depth + 1, right_view)